### PR TITLE
Support as input and use pathlib everywhere

### PR DESCRIPTION
**Pull request by jains8844**
### Description

Fixes #<Issue Number>

In this I did few changes to following files : 
1. sunpy/io/ana.py
2. sunpy/io/fits.py
3. sunpy/io/jp2.py
4. sunpy/io/file_tools.py
5. sunpy/map/map_factory.py

These are made just to get an overview of how can we use `pathlib` for paths in the project.
I will be doing more things further.

### CODE REVIEW

**Comment by nabobalis** [comment]
Maybe I didn't understand the original issue but I'm sure we still want to support string input for those functions. We also want to support path input as well. It seems from a quick glance that changed it so only paths are supported.

Also there seems to be a lot of empty lines removed. I'm not sure they should be removed.

**Comment by nabobalis** [comment]
Yeah, I would have thought we would need checks like this: https://github.com/sunpy/sunpy/commit/58edb1f15f970ee033938121983a0dd773fe02b7#diff-d15d137a3a143cd887b553eb0cdc0376R340 for the functions.

File: sunpy/map/map_factory.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 import glob
 from collections import OrderedDict
 import warnings
+from pathlib import *
```
Comment (vn-ki) [inline sunpy/map/map_factory.py]: I think you meant 

```
import pathlib
```

File: sunpy/map/map_factory.py [hunk 2]
```diff
@@ -53,33 +54,22 @@ class DatabaseEntry(object):
 class MapFactory(BasicRegistrationFactory):
     """
     Map(\*args, \*\*kwargs)
-
```
Comment (vn-ki) [inline sunpy/map/map_factory.py]: Why were all these lines removed?

File: sunpy/io/ana.py [hunk 3]
```diff
@@ -36,31 +34,27 @@ def read(filename, debug=False, **kwargs):
     """
     Loads an ANA file and returns the data and a header in a list of (data,
     header) tuples.
-
     Parameters
     ----------
     filename : `str`
```
Comment (vn-ki) [inline sunpy/io/ana.py]: Should be `str`, `pathlib.Path`

File: sunpy/io/file_tools.py [hunk 4]
```diff
@@ -192,8 +175,8 @@ def _detect_filetype(filepath):
     # FITS
     #
     # Check the extensions to see if it is a gzipped FITS file
-    filepath_rest_ext1, ext1 = os.path.splitext(filepath)
-    _, ext2 = os.path.splitext(filepath_rest_ext1)
+    filepath_rest_ext1, ext1 = filepath.stem,filepath.suffix
+    _, ext2 = filepath_rest_ext1.suffix
```
Comment (vn-ki) [inline sunpy/io/file_tools.py]: `filepath_rest_ext1` is a string after line 178, right?

File: sunpy/io/file_tools.py [hunk 5]
```diff
@@ -169,20 +154,18 @@ def _detect_filetype(filepath):
...
-
     Returns
     -------
     filetype : `str`
         The type of file.
     """
 
     # Open file and read in first two lines
-    with open(filepath, 'rb') as fp:
+    with open(str(filepath), 'rb') as fp:
```
Comment (vn-ki) [inline sunpy/io/file_tools.py]: `open` can take a `Path` type. No need to cast it to `str`.

File: sunpy/io/file_tools.py [hunk 6]
```diff
@@ -52,25 +53,20 @@ def __getitem__(self, key):
 def read_file(filepath, filetype=None, **kwargs):
     """
     Automatically determine the filetype and read the file.
-
     Parameters
     ----------
-    filepath : `str`
+    filepath : `PurePath Object`
```
Comment (vn-ki) [inline sunpy/io/file_tools.py]: Change to `pathlib.PurePath`

File: sunpy/io/file_tools.py [hunk 7]
```diff
@@ -92,22 +88,16 @@ def read_file(filepath, filetype=None, **kwargs):
...
     """
     Reads the header from a given file.
-
     This should always return a instance of io.header.FileHeader
-
     Parameters
     ----------
-
-    filepath : `str`
+    filepath : `PurePath object`
```
Comment (vn-ki) [inline sunpy/io/file_tools.py]: Same as above.

File: sunpy/io/file_tools.py [hunk 8]
```diff
@@ -169,20 +154,18 @@ def _detect_filetype(filepath):
     """
     Attempts to determine the type of data contained in a file.  This is only
     used for reading because it opens the file to check the data.
-
     Parameters
     ----------
-    filepath : `str`
+    filepath : `PurePosixPath/PosixPath`
```
Comment (vn-ki) [inline sunpy/io/file_tools.py]: `PosixPath` and `PurePosixPath` are for use on non-windows systems only. Why is this `PosixPath`?

File: sunpy/io/fits.py [hunk 9]
```diff
@@ -103,22 +95,20 @@ def get_header(afile):
...
     afile : `str` or fits.HDUList
         The file to be read, or HDUList to process.
-
     Returns
     -------
     headers : `list`
         A list of FileHeader headers.
     """
-    if isinstance(afile, fits.HDUList):
+    if isinstance(str(afile), fits.HDUList):
```
Comment (vn-ki) [inline sunpy/io/fits.py]: Why are you casting `afile` into `str` and whether it's an instance of `fits.HDUList`?

File: sunpy/io/fits.py [hunk 10]
```diff
@@ -52,27 +47,24 @@
...
-
     Notes
     -----
     This routine reads all the HDU's in a fits file and returns a list of the
     data and a FileHeader instance for each one.
     Also all comments in the original file are concatenated into a single
     'comment' key in the returned FileHeader.
     """
-    with fits.open(filepath, ignore_blank=True, memmap=memmap) as hdulist:
+    with fits.open(str(filepath), ignore_blank=True, memmap=memmap) as hdulist:
```
Comment (vn-ki) [inline sunpy/io/fits.py]: `fits.open` should support `pathlib`

**Comment by nabobalis** [comment]
Fork has been deleted, so I am closing this.

**Pull request closed without merging**